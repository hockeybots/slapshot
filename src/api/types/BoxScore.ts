/**
 * @typedef {object} Game
 * @property {number} id The ID of the game.
 * @description A representation of an NHL Game.
 */

interface BoxScore {
  id: number;
}

export default BoxScore;
