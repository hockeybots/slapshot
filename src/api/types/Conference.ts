/**
 * @typedef {object} Conference
 * @property {string} abbreviation - The abbreviation for the conference.
 * @property {active} active - Whether the conference is active or not.
 * @property {number} id - ID for the conference as returned by the NHL API.
 * @property {string} name - The name of the conference.
 * @property {string} shortName - The short name for the conference (EG: West for Western).
 * @description A representation of an NHL Conference.
 */
interface Conference {
  abbreviation: string;
  active: boolean;
  id: number;
  name: string;
  shortName: string;
}

export default Conference;
