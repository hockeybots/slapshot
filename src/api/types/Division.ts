/**
 * @typedef {object} Division
 * @property {string} abbreviation The divisions abbreviated name
 * @property {string} name The name of the division.
 * @description A representation of an NHL Division.
 */
interface Division {
  abbreviation: string;
  name: string;
}

export default Division;
