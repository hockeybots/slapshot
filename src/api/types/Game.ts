export default interface Game {
  id: number;
}
