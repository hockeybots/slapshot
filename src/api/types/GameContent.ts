/**
 * @typedef {object} Game
 * @property {number} id The ID of the game.
 * @description A representation of an NHL Game.
 */

interface GameContent {
  id: number;
}

export default GameContent;
