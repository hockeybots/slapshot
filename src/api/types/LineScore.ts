/**
 * @typedef {object} Game
 * @property {number} id The ID of the game.
 * @description A representation of an NHL Game.
 */

interface Team {
  id: number;
}

export default Team;
