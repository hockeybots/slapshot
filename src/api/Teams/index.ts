import idx from 'idx';

import { TEAMS_ENDPOINT } from '../../endpoints';
import { Conference, Division, Game, Team, Venue } from '../types';

import Endpoint from '../Endpoint';
import People from '../People';

/**
 * Teams endpoint wrapper
 * @extends Endpoint
 * @description An abstraction of the {TEAMS_ENDPOINT} with a fluent API
 */
class Teams extends Endpoint {
  /**
   * This method will transform API data in to a Conference object.
   * @param apiData {object} - The conference object of the response from the NHL API
   * @returns {Conference}
   */
  public static toConference(apiData: any): Conference {
    return {
      abbreviation: apiData.abbreviation,
      active: apiData.active,
      id: apiData.id,
      name: apiData.name,
      shortName: apiData.shortName,
    };
  }
  /**
   * This method will transform API data in to a Division object.
   * @param apiData {object} - The division object of the response from the NHL API
   * @returns {Division}
   */
  public static toDivision(apiData: any): Division {
    // TODO: Change conference to optional and fetch conference from endpoint if withConference.
    return {
      abbreviation: apiData.abbreviation,
      active: apiData.active,
      conference: this.toConference(apiData.conference),
      id: apiData.id,
      name: apiData.name,
      shortName: apiData.shortName,
    };
  }
  /**
   * This method will transform API data in to a Venue object.
   * @param apiData {object} - The venue object of the response from the NHL API
   * @returns {Venue}
   */
  public static toVenue(apiData: any): Venue {
    return {
      city: apiData.city,
      name: apiData.name,
      timeZoneAbbreviation: apiData.timeZone.tz,
      timeZoneName: apiData.timeZone.id,
      utcOffset: apiData.timeZone.offset,
    };
  }
  /**
   * This method will transform API data in to a Team object.
   * @param apiData {object} - The Team object of the response from the NHL API
   * @param roster {boolean} - Used to determine whether or not to parse roster data
   * @param previousGame {boolean} - Used to determine whether or not to parse previous game data
   * @param nextGame {boolean} - Used to determine whether or not to parse next game data
   * @param stats {boolean} - Used to determine whether or not to parse stats data
   * @returns {Team}
   */
  public static async toTeam(
    apiData: any,
    roster: boolean,
    previousGame: boolean,
    nextGame: boolean,
    stats: boolean,
  ): Promise<Team> {
    const team: Team = {
      abbreviation: apiData.abbreviation,
      active: apiData.active,
      conference: Teams.toConference(apiData.conference),
      division: Teams.toDivision(apiData.division),
      firstYearOfPlay: apiData.firstYearOfPlay,
      id: apiData.id,
      locationName: apiData.locationName,
      name: apiData.name,
      shortName: apiData.shortName,
      siteUrl: apiData.officialSiteUrl,
      teamName: apiData.teamName,
      venue: Teams.toVenue(apiData.venue),
    };
    if (roster) {
      const rosterIds = idx(apiData, (_) => _.roster.roster.map((rosterMember: any) => rosterMember.person.id));
      if (rosterIds && Array.isArray(rosterIds)) {
        team.roster = await new People(...rosterIds).data();
      }
    }
    if (previousGame) {
      const rosterIds = idx(apiData, (_) => _.roster.roster.map((rosterMember: any) => rosterMember.person.id));
    }
    return team;
  }

  private roster: boolean = false;
  private previousGame: boolean = false;
  private nextGame: boolean = false;
  private stats: boolean = false;

  private teamIds: Array<number>;
  /**
   * @constructor
   * @param {number[]} ids Target team id's
   */
  constructor(...ids: Array<number>) {
    super();
    this.teamIds = ids;
    this.initUri();
  }

  public clear() {
    this.initUri();
  }
  public initUri() {
    this.uri = `${TEAMS_ENDPOINT}?teamId=${this.teamIds.join(',')}`;
  }

  /**
   * @description This method is used after building of the URI is complete. It will fetch and parse the NHL API data.
   * @returns {Teams[]}
   */
  public async data(): Promise<Array<Team>> {
    try {
      const apiData = await this.load();
      return this.parseData(apiData);
    } catch (error) {
      return Promise.reject(error);
    }
  }

  /**
   * @description This method will add the necessary query string to the URI for including the roster data.
   * When parsing the API data we map all people ids from the roster to make a call to the
   * PEOPLE_ENDPOINT for rich data.
   * @returns {Teams}
   */
  public withRoster(): this {
    if (!this.roster) {
      this.roster = true;
      this.uri = `${this.uri}&expand=team.roster`;
    }
    return this;
  }
  /**
   * @description This method will add the necessary query string to the URI for including the target team(s)
   * previous game(s) data.
   * @returns {Teams}
   */
  public withPreviousGame(): this {
    if (!this.previousGame) {
      this.previousGame = true;
      this.uri = `${this.uri}&expand=team.schedule.previous`;
    }
    return this;
  }
  /**
   * @description This method will add the necessary query string to the URI for including the target team(s)
   * next game(s) data.
   * @returns {Teams}
   */
  public withNextGame(): this {
    if (!this.nextGame) {
      this.nextGame = true;
      this.uri = `${this.uri}&expand=team.schedule.next`;
    }
    return this;
  }
  /**
   * @description This method will add the necessary query string to the URI for including the target team(s)
   * statistical data.
   * @returns {Teams}
   */
  public withStats(): this {
    if (!this.stats) {
      this.stats = true;
      this.uri = `${this.uri}&expand=team.stats`;
    }
    return this;
  }
  /**
   * @description This method will add all possible query strings to the URI.
   * @returns {Teams}
   */
  public all(): this {
    return this.withRoster()
      .withStats()
      .withPreviousGame()
      .withNextGame();
  }
  /**
   * @description This method will parse the raw NHL API data in to an array of Team objects.
   * @param {object} apiData The raw NHL API data
   * @returns {Team[]}
   */
  public async parseData(apiData: any): Promise<Array<Team>> {
    const teams = idx(apiData, (_) => _.teams);
    if (!teams || !Array.isArray(teams)) {
      return Promise.reject('Unable to parse, missing data');
    }
    return Promise.all<Team>(
      teams.map((team: any) => Teams.toTeam(team, this.roster, this.previousGame, this.nextGame, this.stats)),
    );
  }
}

export default Teams;
