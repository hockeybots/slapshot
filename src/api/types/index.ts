import Conference from './Conference';
import Division from './Division';
import Player from './Player';
import Team from './Team';
import Venue from './Venue';

export { Conference, Division, Player, Team, Venue };
