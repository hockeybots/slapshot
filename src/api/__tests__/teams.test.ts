import Teams from '../Teams';
import { Team } from '../types';

describe('Api', () => {
  describe('Teams', () => {
    describe('uri', () => {
      describe('when not expanding', () => {
        describe('when requesting one team', () => {
          const team = new Teams(1);

          it('should return the correct uri', () => {
            expect(team.uri).toBe('https://statsapi.web.nhl.com/api/v1/teams?teamId=1');
          });
        });
        describe('when requesting mutlitple teams', () => {
          const team = new Teams(1, 2, 3, 4);

          it('should return the correct uri', () => {
            expect(team.uri).toBe('https://statsapi.web.nhl.com/api/v1/teams?teamId=1,2,3,4');
          });
        });
      });
      describe('when expanding', () => {
        describe('withRoster', () => {
          const team = new Teams(1).withRoster();

          // Assure no duplication
          team
            .withRoster()
            .withRoster()
            .withRoster()
            .withRoster()
            .withRoster()
            .withRoster();

          it('should return the correct uri', () => {
            expect(team.uri).toBe('https://statsapi.web.nhl.com/api/v1/teams?teamId=1&expand=team.roster');
          });
        });
        describe('withPreviousGame', () => {
          const team = new Teams(1).withPreviousGame();

          // Assure no duplication
          team
            .withPreviousGame()
            .withPreviousGame()
            .withPreviousGame()
            .withPreviousGame()
            .withPreviousGame()
            .withPreviousGame();

          it('should return the correct uri', () => {
            expect(team.uri).toBe('https://statsapi.web.nhl.com/api/v1/teams?teamId=1&expand=team.schedule.previous');
          });
        });
        describe('withNextGame', () => {
          const team = new Teams(1).withNextGame();

          // Assure no duplication
          team
            .withNextGame()
            .withNextGame()
            .withNextGame()
            .withNextGame()
            .withNextGame()
            .withNextGame();

          it('should return the correct uri', () => {
            expect(team.uri).toBe('https://statsapi.web.nhl.com/api/v1/teams?teamId=1&expand=team.schedule.next');
          });
        });
        describe('withStats', () => {
          const team = new Teams(1).withStats();

          // Assure no duplication
          team
            .withStats()
            .withStats()
            .withStats()
            .withStats()
            .withStats()
            .withStats();

          it('should return the correct uri', () => {
            expect(team.uri).toBe('https://statsapi.web.nhl.com/api/v1/teams?teamId=1&expand=team.stats');
          });
        });
        describe('all', () => {
          const team = new Teams(1).all();

          // Assure no duplication
          team
            .all()
            .all()
            .all()
            .all()
            .all()
            .all();

          it('should return the correct uri', () => {
            expect(team.uri).toBe(
              `https://statsapi.web.nhl.com/api/v1/teams?teamId=1&expand=team.roster&expand=team.stats&expand=team.schedule.previous&expand=team.schedule.next`,
            );
          });
        });
        describe('clear', () => {
          const team = new Teams(1).all();

          // Assure no duplication
          team
            .all()
            .all()
            .all()
            .all()
            .all()
            .all();

          team.clear();

          it('should return the correct uri', () => {
            expect(team.uri).toBe(`https://statsapi.web.nhl.com/api/v1/teams?teamId=1`);
          });
        });
      });
    });
    describe('toConference', () => {
      const mockApiData = {
        abbreviation: 'WW',
        active: true,
        id: 1,
        name: 'Wild West',
        shortName: 'Wild',
      };
      it('should return a conference object with the correct abbreviation', () => {
        expect(Teams.toConference(mockApiData).abbreviation).toBe(mockApiData.abbreviation);
      });
      it('should return a conference object with the correct active status', () => {
        expect(Teams.toConference(mockApiData).active).toBe(mockApiData.active);
      });
      it('should return a conference object with the correct id', () => {
        expect(Teams.toConference(mockApiData).id).toBe(mockApiData.id);
      });
      it('should return a conference object with the correct name', () => {
        expect(Teams.toConference(mockApiData).name).toBe(mockApiData.name);
      });
      it('should return a conference object with the correct short name', () => {
        expect(Teams.toConference(mockApiData).shortName).toBe(mockApiData.shortName);
      });
    });
    describe('toDivision', () => {
      const mockApiData = {
        abbreviation: 'GB',
        active: true,
        id: 5,
        name: 'Gummy Bears',
        nameShort: 'Gummy',
      };
      it('should return a division object with the correct abbreviation', () => {
        expect(Teams.toDivision(mockApiData).abbreviation).toBe(mockApiData.abbreviation);
      });
      it('should return a division object with the correct active status', () => {
        expect(Teams.toDivision(mockApiData).active).toBe(mockApiData.active);
      });
      it('should return a division object with the correct id', () => {
        expect(Teams.toDivision(mockApiData).id).toBe(mockApiData.id);
      });
      it('should return a division object with the correct name', () => {
        expect(Teams.toDivision(mockApiData).name).toBe(mockApiData.name);
      });
      it('should return a division object with the correct short name', () => {
        expect(Teams.toDivision(mockApiData).nameShort).toBe(mockApiData.nameShort);
      });
    });
    describe('toVenue', () => {
      const mockApiData = {
        city: 'Tokyooo',
        name: 'Battle Dome',
        timeZone: {
          tz: 'TST',
          offset: 666,
          id: 'Tokyo Battle Time',
        },
      };
      it('should return a venue object with the correct city', () => {
        expect(Teams.toVenue(mockApiData).city).toBe(mockApiData.city);
      });
      it('should return a venue object with the correct name', () => {
        expect(Teams.toVenue(mockApiData).name).toBe(mockApiData.name);
      });
      it('should return a venue object with the correct timeZoneName', () => {
        expect(Teams.toVenue(mockApiData).timeZoneName).toBe(mockApiData.timeZone.id);
      });
      it('should return a venue object with the correct timeZoneAbbreviation', () => {
        expect(Teams.toVenue(mockApiData).timeZoneAbbreviation).toBe(mockApiData.timeZone.tz);
      });
      it('should return a venue object with the correct utc offset', () => {
        expect(Teams.toVenue(mockApiData).utcOffset).toBe(mockApiData.timeZone.offset);
      });
    });
    describe('toTeam', () => {
      const mockApiData = {
        abbreviation: 'RZR',
        active: true,
        firstYearOfPlay: 1999,
        id: 80085,
        locationName: 'Battle Dome Zone',
        name: 'Wreck City Razors',
        shortName: 'Razors',
        officialSiteUrl: 'https://getwrecked.co.jp',
        teamName: 'Razors',
        division: {
          abbreviation: 'GB',
          active: true,
          id: 5,
          name: 'Gummy Bears',
          nameShort: 'Gummy',
        },
        conference: {
          abbreviation: 'WW',
          active: true,
          id: 1,
          name: 'Wild West',
          shortName: 'Wild',
        },
        venue: {
          city: 'Tokyooo',
          name: 'Battle Dome',
          timeZone: {
            tz: 'TST',
            offset: 666,
            id: 'Tokyo Battle Time',
          },
        },
      };
      it('should return a team object with the correct data', async () => {
        const team = await Teams.toTeam(mockApiData, false, false, false, false);
        expect(team).toMatchSnapshot();
      });
    });
    describe('parseData', () => {
      describe('when data IS NOT passed', () => {
        it('handles the error', () => {
          const player = new Teams(1);
          expect(
            player.parseData<Team>(null, 'teams', (apiDataItem) =>
              Teams.toTeam(apiDataItem, false, false, false, false),
            ),
          ).rejects.toBe('Unable to parse, missing data');
        });
      });
      describe('when data IS passed', () => {
        it('returns an array of Team objects', () => {
          const player = new Teams(1);
          const mockApiData = {
            copyright:
              'NHL and the NHL Shield are registered trademarks of the National Hockey League. NHL and NHL team marks are the property of the NHL and its teams. Â© NHL 2018. All Rights Reserved.',
            teams: [
              {
                id: 1,
                name: 'New Jersey Devils',
                link: '/api/v1/teams/1',
                venue: {
                  name: 'Prudential Center',
                  link: '/api/v1/venues/null',
                  city: 'Newark',
                  timeZone: {
                    id: 'America/New_York',
                    offset: -5,
                    tz: 'EST',
                  },
                },
                abbreviation: 'NJD',
                teamName: 'Devils',
                locationName: 'New Jersey',
                firstYearOfPlay: '1982',
                division: {
                  id: 18,
                  name: 'Metropolitan',
                  nameShort: 'Metro',
                  link: '/api/v1/divisions/18',
                  abbreviation: 'M',
                },
                conference: {
                  id: 6,
                  name: 'Eastern',
                  link: '/api/v1/conferences/6',
                },
                franchise: {
                  franchiseId: 23,
                  teamName: 'Devils',
                  link: '/api/v1/franchises/23',
                },
                shortName: 'New Jersey',
                officialSiteUrl: 'http://www.newjerseydevils.com/',
                franchiseId: 23,
                active: true,
              },
            ],
          };

          const expectedOutput: Array<Team> = [
            {
              abbreviation: 'NJD',
              active: true,
              firstYearOfPlay: '1982',
              id: 1,
              locationName: 'New Jersey',
              name: 'New Jersey Devils',
              shortName: 'New Jersey',
              siteUrl: 'http://www.newjerseydevils.com/',
              teamName: 'Devils',
              division: {
                id: 18,
                name: 'Metropolitan',
                nameShort: 'Metro',
                abbreviation: 'M',
                active: true,
              },
              venue: {
                city: 'Newark',
                name: 'Prudential Center',
                timeZoneAbbreviation: 'EST',
                timeZoneName: 'America/New_York',
                utcOffset: -5,
              },
            },
          ];

          expect(
            player.parseData<Team>(mockApiData, 'teams', (team) => Teams.toTeam(team, false, false, false, false)),
          ).resolves.toEqual(expectedOutput);
        });
      });
    });
  });
});
