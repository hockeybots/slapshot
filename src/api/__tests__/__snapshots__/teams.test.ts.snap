// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Api Teams toTeam should return a team object with the correct data 1`] = `
Object {
  "abbreviation": "RZR",
  "active": true,
  "division": Object {
    "abbreviation": "GB",
    "active": true,
    "id": 5,
    "name": "Gummy Bears",
    "shortName": "Gummy",
  },
  "firstYearOfPlay": 1999,
  "id": 80085,
  "locationName": "Battle Dome Zone",
  "name": "Wreck City Razors",
  "shortName": "Razors",
  "siteUrl": "https://getwrecked.co.jp",
  "teamName": "Razors",
  "venue": Object {
    "city": "Tokyooo",
    "name": "Battle Dome",
    "timeZoneAbbreviation": "TST",
    "timeZoneName": "Tokyo Battle Time",
    "utcOffset": 666,
  },
}
`;
